aiohttp/payload.py

```diff
@@ -328,15 +400,165 @@ def __init__(
+
+        Args:
+            remaining_content_len: Optional maximum number of bytes to read.
+                If None, READ_SIZE will be used as the default chunk size.
+
+        Returns:
+            A chunk of bytes read from the file object, respecting the
+            remaining_content_len limit if specified.
+
+        This method is used for subsequent reads during streaming after
+        the initial _read_and_available_len call has been made.
+
+        """
+        return self._value.read(remaining_content_len or READ_SIZE)  # type: ignore[no-any-return]
+
+    @property
+    def size(self) -> Optional[int]:
+        try:
+            return os.fstat(self._value.fileno()).st_size - self._value.tell()
+        except (AttributeError, OSError):
+            return None
+
     async def write(self, writer: AbstractStreamWriter) -> None:
-        loop = asyncio.get_event_loop()
+        """
+        Write the entire file-like payload to the writer stream.
+
+        Args:
+            writer: An AbstractStreamWriter instance that handles the actual writing
+
+        This method writes the entire file content without any length constraint.
+        It delegates to write_with_length() with no length limit for implementation
+        consistency.
+
+        Note:
+            For new implementations that need length control, use write_with_length() directly.
+            This method is maintained for backwards compatibility with existing code.
+
+        """
+        await self.write_with_length(writer, None)
+
+    async def write_with_length(
+        self, writer: AbstractStreamWriter, content_length: Optional[int]
+    ) -> None:
+        """
+        Write file-like payload with a specific content length constraint.
+
+        Args:
+            writer: An AbstractStreamWriter instance that handles the actual writing
+            content_length: Maximum number of bytes to write (None for unlimited)
+
+        This method implements optimized streaming of file content with length constraints:
+
+        1. File reading is performed in a thread pool to avoid blocking the event loop
+        2. Content is read and written in chunks to maintain memory efficiency
+        3. Writing stops when either:
+           - All available file content has been written (when size is known)
+           - The specified content_length has been reached
+        4. File resources are properly closed even if the operation is cancelled
+
+        The implementation carefully handles both known-size and unknown-size payloads,
+        as well as constrained and unconstrained content lengths.
+
+        """
+        loop = asyncio.get_running_loop()
+        total_written_len = 0
+        remaining_content_len = content_length
+
         try:
-            chunk = await loop.run_in_executor(None, self._value.read, 2**16)
+            # Get initial data and available length
+            available_len, chunk = await loop.run_in_executor(
```

Comment:
Should we just use asyncio.to_thread() now?

Reply:
```
async def to_thread(func, /, *args, **kwargs):
    """Asynchronously run function *func* in a separate thread.

    Any *args and **kwargs supplied for this function are directly passed
    to *func*. Also, the current :class:`contextvars.Context` is propagated,
    allowing context variables from the main thread to be accessed in the
    separate thread.

    Return a coroutine that can be awaited to get the eventual result of *func*.
    """
    loop = events.get_running_loop()
    ctx = contextvars.copy_context()
    func_call = functools.partial(ctx.run, func, *args, **kwargs)
    return await loop.run_in_executor(None, func_call)
```

Well we still need `loop.run_in_executor` for the `_schedule_file_close` so it probably makes sense to leave it as-is for now.